src/Query/QueryBuilder.php

```diff
@@ -302,14 +309,47 @@ public function fetchFirstColumn(): array
      */
     public function executeQuery(): Result
     {
+        [$params, $types] = $this->buildParametersAndTypes();
+
         return $this->connection->executeQuery(
             $this->getSQL(),
-            $this->params,
-            $this->types,
+            $params,
+            $types,
             $this->resultCacheProfile,
         );
     }
 
+    /**
+     * Build then return parameters and types for the query.
+     *
+     * @return array{
+     *     list<mixed>|array<string, mixed>,
+     *     WrapperParameterTypeArray,
+     * } The parameters and types for the query.
+     */
+    private function buildParametersAndTypes(): array
+    {
+        $cteParams = $cteParamTypes = [];
+
+        foreach ($this->commonTableExpressions as $cte) {
+            if (! $cte->query instanceof self) {
+                continue;
+            }
+
+            $cteParams     = array_merge($cteParams, $cte->query->params);
```

Comment:
This overrides named placehoders, if two part QueryBuilder instances auto creates named placeholders, specially when `QueryBuilder->createNamedParameter()` are used with automatic name generation based on a counter:

```php
        $cteQueryBuilder1 = $this->connection->createQueryBuilder();
        $cteQueryBuilder1->select('id AS virtual_id')
            ->from('for_update')
            ->where('virtual_id = ' . $cteQueryBuilder1->createNamedParameter(123, ParameterType::INTEGER));

        $cteQueryBuilder2 = $this->connection->createQueryBuilder();
        $cteQueryBuilder2->select('id AS virtual_id')
            ->from('cte_a')
            ->where('virtual_id = ' . $cteQueryBuilder2->createNamedParameter(987, ParameterType::INTEGER));
```

would merge them in a weired way. Additionally, the tests related to named placeholder only covers manual defining them with setting parameter, not using the create methods along with autogeneration and using in two sub-querybuilder instances which gets merged.

https://github.com/doctrine/dbal/blob/c91166a91b8639a1d7bfe0f88bca0273bdbe9d2d/src/Query/QueryBuilder.php#L1403-L1440

Tests related to the `placeholder` merging and usage is not sufficient enough.

Reply:
That does not work with default (auto) placeholder because of the `boundCounter` value which start to 1 for each CTE.
But it is ok when setting placeholder
```php
->where($cteQueryBuilder1->expr()->eq(
    'id',
    $cteQueryBuilder1->createNamedParameter(1, ParameterType::INTEGER, ':id1'),
));

->where($cteQueryBuilder2->expr()->in(
    'id',
    $cteQueryBuilder2->createNamedParameter([1, 2], ArrayParameterType::INTEGER, ':id2'),
));     
```
I've added some tests for this

Reply:
My whole point is, that starting this leads to more things some developers **needs** to know beside a simple "use the top QueryBuilder" instance. Not a fan of starting this and leave it half way broken and unfinished. But that is more a personal taste:

```
->where($cteQueryBuilder1->expr()->eq(
    'id',
    $cteQueryBuilder1->createNamedParameter(1, ParameterType::INTEGER, ':id'),
));

->where($cteQueryBuilder2->expr()->in(
    'id',
    $cteQueryBuilder2->createNamedParameter([1, 2], ArrayParameterType::INTEGER, ':id'),
));  
```

In that case database error should occor and be a little bit of help that something is off when merging the CTE parts on query execution.

```
->where($cteQueryBuilder1->expr()->eq(
    'id',
    $cteQueryBuilder1->createNamedParameter(1, ParameterType::INTEGER, ':id'),
));

->where($cteQueryBuilder2->expr()->in(
    'id',
    $cteQueryBuilder2->createNamedParameter(2, ParameterType::INTEGER, ':id'),
));  
```

would merge that and use `2` for both places for the same `:id` without any notices - and will give a hard time to debug and find this.

Askin me the whole merging should vanish again until it could be implemented completly in all parts and with a concept behind it. But that is not my decision.

Let's see what @derrabus would say.

Reply:
I agree, it could give a hard time to debug this. We could prevent this by adding a check during the process and reject a parameter's name that is already used.